pkg/security/ebpf/table.go

```diff
@@ -0,0 +1,131 @@
+// +build linux
+
+package ebpf
+
+import (
+	"bytes"
+	"encoding/binary"
+	"unsafe"
+
+	bpflib "github.com/iovisor/gobpf/elf"
+)
+
+// Table represents an eBPF map
+type Table struct {
+	*bpflib.Map
+	module *bpflib.Module
+}
+
+// TableItem represents an eBPF map item, either a key or a vlue
+type TableItem interface {
+	Bytes() ([]byte, error)
+}
+
+// Get retrieves the value associated with a specified key
+func (t *Table) Get(key []byte) ([]byte, error) {
+	var value [1024]byte
+	err := t.module.LookupElement(t.Map, unsafe.Pointer(&key[0]), unsafe.Pointer(&value[0]))
+	return value[:], err
+}
+
+// Set associates a value with a key in an eBPF map
+func (t *Table) Set(key, value TableItem) error {
+	bKey, err := key.Bytes()
+	if err != nil {
+		return err
+	}
+	bValue, err := value.Bytes()
+	if err != nil {
+		return err
+	}
+	return t.module.UpdateElement(t.Map, unsafe.Pointer(&bKey[0]), unsafe.Pointer(&bValue[0]), 0)
+}
+
+// SetP associates a value with a key in an eBPF map
+func (t *Table) SetP(key, value []byte) error {
+	return t.module.UpdateElement(t.Map, unsafe.Pointer(&key[0]), unsafe.Pointer(&value[0]), 0)
+}
+
+// GetNext returns the key following the one passed as parameter
+func (t *Table) GetNext(key []byte) (bool, []byte, []byte, error) {
+	var value [1024]byte
+	pKey := unsafe.Pointer(&key[0])
+	nextKey := make([]byte, len(key))
+	more, err := t.module.LookupNextElement(t.Map, pKey, unsafe.Pointer(&nextKey[0]), unsafe.Pointer(&value[0]))
+	return more, nextKey, value[:], err
+}
+
+// Delete the entry associated with the specified key
+func (t *Table) Delete(key []byte) error {
+	return t.module.DeleteElement(t.Map, unsafe.Pointer(&key[0]))
+}
+
+// BytesTableItem describes a raw table key or value
+type BytesTableItem []byte
+
+// Bytes returns the binary representation of a BytesTableItem
+func (i BytesTableItem) Bytes() ([]byte, error) {
+	return []byte(i), nil
+}
+
+// Uint8TableItem describes an uint8 table key or value
+type Uint8TableItem uint8
+
+// Bytes returns the binary representation of a Uint8TableItem
+func (i Uint8TableItem) Bytes() ([]byte, error) {
+	return []byte{uint8(i)}, nil
+}
+
+// Uint32TableItem describes an uint32 table key or value
+type Uint32TableItem uint32
+
+// Bytes returns the binary representation of a Uint32TableItem
+func (i Uint32TableItem) Bytes() ([]byte, error) {
+	b := make([]byte, 4)
+	ByteOrder.PutUint32(b, uint32(i))
+	return b, nil
+}
+
+// Uint64TableItem describes an uint64 table key or value
+type Uint64TableItem uint64
+
+// Bytes returns the binary representation of a Uint64TableItem
+func (i Uint64TableItem) Bytes() ([]byte, error) {
+	b := make([]byte, 8)
+	ByteOrder.PutUint64(b, uint64(i))
+	return b, nil
+}
+
+// StringTableItem describes an string table key or value
+type StringTableItem struct {
+	str  string
```

Comment:
Should fields be exported in this exported struct? Or should it be implementing an exported interface and be private to the package?

Reply:
It is implementing the `TableItem` interface and can be instantiated using `NewStringTableItem`. What would be the value of exporting the fields ?

Reply:
Thank you. This is a small nit really, but if you can't initialize a struct outside of this package, then maybe it shouldn't be exported? Are there currently use cases that require this type, e.g. in a `switch` on `.(type)`, mind pointing me at them? If they are only used via the implemented interface then maybe keep this and other types private to the package?

Reply:
There is a constructor for this `NewStringTableItem`

Reply:
Yes, but do you need to know about this type `StringTableItem` outside of this package, you just use it via the interface? Why export this type and other types implementing the `TableItem`. E.g. here: https://github.com/DataDog/datadog-agent/pull/6014/files#diff-55f7608a287af8fc4223cf27806e86b3R186. Am I missing some other use case?

Reply:
When you look at the generated doc, it's clearer to see the type and its description and not just the constructor.